modules/apigee_edge_debug/src/HttpClientMiddleware/DevelKintApiClientProfiler.php

```diff
@@ -0,0 +1,165 @@
+use GuzzleHttp\TransferStats;
+use Psr\Http\Message\RequestInterface;
+use Psr\Log\LogLevel;
+
+/**
+ * Http client middleware that profiles Apigee Edge API calls.
+ */
+class DevelKintApiClientProfiler {
+
+  /**
+   * The currently logged-in user.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  private $currentUser;
+
+  /**
+   * The messenger service.
+   *
+   * @var \Drupal\Core\Messenger\MessengerInterface
+   */
+  private $messenger;
+
+  /**
+   * The debug message formatter plugin.
+   *
+   * @var \Drupal\apigee_edge_debug\Plugin\DebugMessageFormatter\DebugMessageFormatterPluginInterface
+   */
+  private $formatter;
+
+  /**
+   * The module handler.
+   *
+   * @var \Drupal\Core\Extension\ModuleHandlerInterface|null
+   */
+  private $moduleHandler;
+
+  /**
+   * DevelKintApiClientProfiler constructor.
+   *
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   Config factory.
+   * @param \Drupal\apigee_edge_debug\DebugMessageFormatterPluginManager $debug_message_formatter_plugin
+   *   Debug message formatter plugin manager.
+   * @param \Drupal\Core\Session\AccountInterface $currentUser
+   *   The currently logged-in user.
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+   *   The module handler service.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   The messenger service.
+   */
+  public function __construct(ConfigFactoryInterface $config_factory, DebugMessageFormatterPluginManager $debug_message_formatter_plugin, AccountInterface $currentUser, ModuleHandlerInterface $module_handler, MessengerInterface $messenger) {
+    // On module install, this constructor is called earlier than
+    // the module's configuration would have been imported to the database.
+    // In that case the $formatterPluginId is missing and it causes fatal
+    // errors.
+    $formatter_plugin_id = $config_factory->get('apigee_edge_debug.settings')->get('formatter');
+    if ($formatter_plugin_id) {
+      $this->formatter = $debug_message_formatter_plugin->createInstance($formatter_plugin_id);
+    }
+    $this->currentUser = $currentUser;
+    $this->moduleHandler = $module_handler;
+    $this->messenger = $messenger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __invoke() {
+    // If the formatter has been initialized yet then do nothing.
+    if (!$this->formatter) {
+      return function ($handler) {
+        return function (RequestInterface $request, array $options) use ($handler) {
+          return $handler($request, $options);
+        };
+      };
+    }
+    return function ($handler) {
+      return function (RequestInterface $request, array $options) use ($handler) {
+
+        $formatter = $this->formatter;
+
+        $options[RequestOptions::ON_STATS] = function (TransferStats $stats) use ($request, $formatter) {
+          // Do not modify the original request object in the subsequent calls.
```

Comment:
Possible unnecessary code duplications here? I have the check that later...

Reply:
@mxr576 Feel free to be specific about any suggestions for the author of this PR. Thanks